Overview: Command line argument parsing, lexer (#659)
rust-lang/rustc-dev-guide: src/overview.md

**TODO: someone else should confirm this vvv**

> - The compile process begins when a user writes a Rust source program in text and invokes the `rustc` compiler on it. The work that the compiler needs to perform is defined with command line options. For example, it is possible to optionally enable nightly features, perform `check`-only builds, or emit LLVM-IR rather than complete the entire compile process defined here. The `rustc` executable call may be indirect through the use of `cargo`.

Reviewer: Should we have a soft wrap here?

Author: Open question how we will format the entire document.  I am thinking that maybe this will become paragraph-based body copy rather than a bullet list?

Author: @mark-i-m Open issue here if we merge

> - The compile process begins when a user writes a Rust source program in text and invokes the `rustc` compiler on it. The work that the compiler needs to perform is defined with command line options. For example, it is possible to optionally enable nightly features, perform `check`-only builds, or emit LLVM-IR rather than complete the entire compile process defined here. The `rustc` executable call may be indirect through the use of `cargo`.

Reviewer: Few suggestions:
```suggestion
- The compile process begins when a user writes a Rust source program in text and invokes the `rustc` compiler on it. The work that the compiler needs to perform is defined by command-line options. For example, it is possible to enable nightly features (`-Z` flags), perform `check`-only builds, or emit LLVM-IR rather than executable machine code. The `rustc` executable call may be indirect through the use of `cargo`.
```

Author: Ty!

> - Command line argument parsing occurs in the [`librustc_driver`]. This crate defines the compile configuration that is requested by the user.

Reviewer: I'm tempted to add what the configuration is turned into:
```suggestion
- Command line argument parsing occurs in the [`librustc_driver`]. This crate defines the compile configuration that is requested by the user and passes it to the rest of the compilation process as a [`rustc_interface::Config`].
```
(`rustc_interface::Config` is https://doc.rust-lang.org/nightly/nightly-rustc/rustc_interface/interface/struct.Config.html)

But I don't know if this too much of a detail to be included in the Overview...

Reviewer 2: am happy with the change... it's not too much

Author: Agree but don't know this area of the code.  Can you confirm that this is the correct reference?  If not, I'll take a look this weekend and commit your suggestion here.

- The raw Rust source text is analyzed by a low-level lexer located in [`librustc_lexer`]. At this stage, the source text is turned into a stream of atomic source code units known as _tokens_. (**TODO**: chrissimpkins - Maybe discuss Unicode handling during this stage?)

> - The token stream passes through a higher-level lexer located in [`librustc_parse`] to prepare for the next stage of the compile process. The [`StringReader`] struct is used at this stage to perform a set of validations and turn strings into interned symbols.

Reviewer: Did we define interning before? If not, maybe include a link to the chapter about interning?

Author: @mark-i-m Open issue here if we merge
